### ISSUE TITLE
Mirage Models

### ISSUE THREAD

**Issue by lcpriest** [samselikoff/ember-cli-mirage#544]
```
As a clarifying point, Mirage models know nothing about your Ember application, including any Ember Data models you may have defined.
```

Is this distinction a goal, or is there going to be some unification? I can't currently see any case when my mirage model would be different than an Ember Data one.

**Comment by samselikoff** [comment]
To answer your first question, this is the reality of Mirage at the moment. Mirage literally knows nothing about your Ember app.

To the second point, if you're following proper conventions there shouldn't be any difference. The main reason I don't piggyback off of ED's models to ascertain relationship information is because Mirage works for Ember apps that don't use Ember Data. Also, there's an eventual goal to extract core Mirage.js from Ember CLI Mirage, and make it available for non-Ember audiences.

For now, defining Mirage models that mirror your ED models is a bit tedious, though in the 0.1.x series, you had to define either a fixture or factory file to get your database collections set up anyway. So, if you had empty factory or fixture files, you won't need those anymore.

The main idea is, models are the canonical source of info about your "database schema." This way, there's a single place where I know to look for your schema info, in addition to giving you a place to define your relationships.

Finally, these changes have all foundation work that pave the way for future ergonomic improvements, notably

 - an improved factory API that lets you specify relationships in your factory definitions, a la factory_girl
 - an `ember-cli-mirage-ember-data` addon, which would read in your ED models and use those to set up Mirage's ORM at run-time

**Comment by lcpriest** [comment]
Thanks for the comprehensive answer. I think I was mentally placing mirage in the same position as FactoryGirl. 

The `ember-cli-mirage-ember-data` addon sounds like it will be very useful.

**Comment by samselikoff** [comment]
You're welcome! Parts of Mirage are definitely heavily inspired by FactoryGirl, but for the *mock server*, rather than your Ember app, if that makes sense. Meaning, first you mock your server, then use a FactoryGirl-like API (soon very limited subset, eventually more robust) to set up the "server state" during your tests. 

There is another addon called Ember Data Factory Guy, which has a powerful factory system for your Ember Data models. Mirage's take is a bit different, since it begins with mocking your server, whereas Factory Guy's approach is, let's set up your local  `store` with the models you need. Different approaches that accomplish roughly the same thing.

**Comment by lcpriest** [comment]
That was where my first assumption totally fell down! 
Side question, the docs for the model reference how to define associations, but it looks like the attributes are pretty freeform; is that correct?

**Comment by samselikoff** [comment]
Yep, it's a "schemaless" db so you don't need to define your attrs. Whatever you pass into a `create` call will get used.

**Comment by lcpriest** [comment]
```
import { Model, belongsTo, hasMany } from 'ember-cli-mirage';

export default Model.extend({
  board: belongsTo(),
  user: belongsTo('creator'),
  section: belongsTo(),
  user: belongsTo('verificationRequester'),
  user: belongsTo('verifier'),
  comments: hasMany(),
});
```

```
import { Model, belongsTo, hasMany } from 'ember-cli-mirage';

export default Model.extend({
  board: belongsTo(),
  creator: belongsTo('user'),
  section: belongsTo(),
  verificationRequester: belongsTo('user'),
  verifier: belongsTo('user'),
  comments: hasMany(),
});
```
Hi again, Sorry for dragging this issue out so much!
Which is the correct syntax? I would have assumed the second, but that throws a 'cannot redefine userId' error.

**Comment by samselikoff** [comment]
This is actually a known issue: https://github.com/samselikoff/ember-cli-mirage/issues/528

Currently the ORM uses the type (`user` in this case) for foreign keys, instead of the property name, so your multiple belongsTo user properties are colliding. This will be fixed soon (it's one of the issues blocking 0.2.0).
